### PR TITLE
[MPS] Fix index_put with deterministic algorithm enabled

### LINKED ISSUE #97574: MPS `index_put_` non-deterministic even with `torch.use_deterministic_algorithm(True)`
### 🐛 Describe the bug

`torch.Tensor.__setitem__` (which according to the documentation is equivalent to calling `torch.Tensor.index_put_`) is non-deterministic even with `torch.use_deterministic_algorithm(True)`. I ran the following code and got non-deterministic outputs (however, occasionally the output is deterministic, so you may have to run it a few times to see the non-determinism.)
```
import torch

torch.use_deterministic_algorithms(True)
device = torch.device('mps')

idx = torch.tensor([
    0, 0, 0, 0, 2, 2, 1, 0, 2, 1, 0, 1, 2, 1, 0, 2, 2, 2, 2, 2, 0, 0, 2, 1, 2, 1, 0, 0, 2, 0, 2, 1, 1, 2, 2, 0, 2, 1, 0, 2
], device=device)

for _ in range(5):
    t = torch.zeros(3, dtype=torch.long, device=device)
    t[idx] = torch.arange(len(idx), device=device)
    print(t)
```
Output:
```
tensor([35, 32, 33], device='mps:0')
tensor([20, 23, 16], device='mps:0')
tensor([35, 32, 33], device='mps:0')
tensor([20, 23, 16], device='mps:0')
tensor([35, 32, 33], device='mps:0')
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230320
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.28)
CMake version: version 3.26.0
Libc version: N/A

Python version: 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:38:11)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.1.0.dev20230320
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230320          pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by qqaatw**
Prevent using parallel computing when deterministic algorithm is set.

Fixes #97574

Benchmark:
```
[--------------- index_put_ Deterministic Algorithm Enabled ---------------]
                                                              |  cpu  |  mps
1 threads: -----------------------------------------------------------------
      Dtype: torch.float32 Features: 1024; Num Indices: 512   |   37  |   49
      Dtype: torch.float32 Features: 1024; Num Indices: 1024  |   54  |   50
      Dtype: torch.float32 Features: 1024; Num Indices: 2048  |   86  |   50
      Dtype: torch.float32 Features: 1024; Num Indices: 4096  |  150  |   49

Times are in microseconds (us).

[-------------- index_put_ Deterministic Algorithm Disabled ---------------]
                                                              |  cpu  |  mps
1 threads: -----------------------------------------------------------------
      DType: torch.float32 Features: 1024; Num Indices: 512   |   37  |   49
      DType: torch.float32 Features: 1024; Num Indices: 1024  |   53  |   49
      DType: torch.float32 Features: 1024; Num Indices: 2048  |   86  |   49
      DType: torch.float32 Features: 1024; Num Indices: 4096  |  147  |   50

Times are in microseconds (us).
```

### <samp>🤖 Generated by Copilot at ebf2ff3</samp>

Added a deterministic version of `index_put` for MPS tensors that runs on a single thread and can be enabled by a global context flag. Refactored the existing `index_put` function and the kernel selection logic to support both parallel and serial modes. Added a test function to verify the deterministic behavior of `index_put` under different conditions.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 1]
```diff
@@ -39,6 +39,7 @@ static bool dispatchIndexKernel(TensorIteratorBase& iter,
   if (iter.numel() == 0) {
     return true;
   }
+  bool serial_index_put = at::globalContext().deterministicAlgorithms() && !accumulate && !index_select;
```
Comment (qqaatw) [inline aten/src/ATen/native/mps/operations/Indexing.mm:42]:
```suggestion
  const bool serial_index_put = at::globalContext().deterministicAlgorithms() && !accumulate && !index_select;
```

**Review by malfet — CHANGES_REQUESTED** [review]
This feels like a huge performance penalty for non-deteriminstic invocations. (Though may be I'm wrong, can you please post some benchmark numbers as part of PR description)
Instead of passing extra argument, perhaps it would be more efficient to schedule it as a single threaded op?)


And how is similar algorithm implemented for CPU?

**Comment by qqaatw** [comment]
> This feels like a huge performance penalty for non-deteriminstic invocations. (Though may be I'm wrong, can you please post some benchmark numbers as part of PR description) Instead of passing extra argument, perhaps it would be more efficient to schedule it as a single threaded op?)
> 
> And how is similar algorithm implemented for CPU?

1. Sure I can benchmark it but why would it result in performance penalty? the previous non-deterministic approach is not changed in this PR. The new deterministic approach is branched on `serial_index_put` condition.
2. Making it a single threaded op sounds reasonable.
3. The CPU impl. also disables parallelism when deterministic is set.

https://github.com/pytorch/pytorch/blob/bb40b62501102be5a23c45b2070b2e7ef3753641/aten/src/ATen/native/cpu/IndexKernel.cpp#L115

**Review by kulinseth — APPROVED** [review]
Thanks @qqaatw for looking into this.
The change does affect performance. Can you please check that the performance didn't regress on the default path. You can run `torchbench` networks to check that.

**Comment by qqaatw** [comment]
> Thanks @qqaatw for looking into this. The change does affect performance. Can you please check that the performance didn't regress on the default path. You can run `torchbench` networks to check that.

Thanks. I replied offline for more detailed discussions.

**Pull request closed without merging**